components/script/dom/bluetoothpermissionresult.rs

```diff
@@ -0,0 +1,219 @@
+        // Step 6.
+        for allowed_device in allowed_devices {
+            // Step 6.1.
+            if let Some(id) = bluetoth_descriptor.deviceId.clone() {
+                if allowed_device.deviceId != id {
+                    continue;
+                } else {
+                    device_id = String::from(id);
+                }
+            }
+
+            // Step 6.2.
+            // Instead of creating an internal slot we send an ipc message to the Bluetooth thread
+            // to check if one of the filters matches.
+            if let Some(ref filters) = bluetoth_descriptor.filters {
+                let mut scan_filters: Vec<BluetoothScanfilter> = Vec::new();
+                // TODO: Create an issue for the spec, to make the canonicalization step here.
+                for filter in filters {
+                    match canonicalize_filter(&filter) {
+                        Ok(f) => scan_filters.push(f),
+                        Err(err) => {
+                            return promise.reject_error(cx, err);
+                        },
+                    }
+                }
+                let (sender, receiver) = ipc::channel().unwrap();
+                self.get_bluetooth_thread()
+                    .send(BluetoothRequest::MatchesFilter(device_id.clone(),
+                                                          BluetoothScanfilterSequence::new(scan_filters),
+                                                          sender))
+                    .unwrap();
+
+                match receiver.recv().unwrap() {
+                    Ok(true) => (),
+                    Ok(false) => continue,
+                    Err(err) => {
+                        return promise.reject_error(cx, Error::from(err));
+                    },
+                };
+            }
+
+            // Step 6.4.
+            // TODO: Implement this correctly, not just using device ids here.
+            // https://webbluetoothcg.github.io/web-bluetooth/#get-the-bluetoothdevice-representing
+            if let Some(ref device) = device_map.get(&device_id) {
+                matching_devices.push(JS::from_ref(&**device.clone()));
+            }
+        }
+
+        // Step 7.
+        *self.devices.borrow_mut() = matching_devices;
+
+        // https://w3c.github.io/permissions/#dom-permissions-query
+        // Step 7.
+        promise.resolve_native(cx, &self.parent);
+    }
+
+    // https://webbluetoothcg.github.io/web-bluetooth/#bluetoothpermissionresult
+    pub fn permission_request(&self, promise: &Rc<Promise>) -> ErrorResult {
+        let descriptor = self.parent.get_query().borrow();
+        let options = match *descriptor {
+            PermissionDescriptorType::Bluetooth(ref d) => d,
+            _ => return Err(Error::Type(DESCRIPTOR_TYPE_ERROR.to_owned())),
+        };
+
+        // Step 1.
+        if (options.filters.is_some() && options.acceptAllDevices) ||
+           (options.filters.is_none() && !options.acceptAllDevices) {
```

Comment:
I think you are half right, because `(a & b) | (~a & ~b)` is a tautology so it can be replaced with another tautology ofc., but the reason why this works against this expression returns always true in logical manner is, that the `&&` and `||` are [lazily evaluated](https://doc.rust-lang.org/reference.html#lazy-boolean-operators).
This logic is already used [here](https://github.com/servo/servo/blob/master/components/script/dom/bluetooth.rs#L445) in our code, and is tested as well, so I want to keep this form.